fish-rust/src/builtins/string/collect.rs

```diff
@@ -0,0 +1,71 @@
+use super::{Arguments, StringError, StringSubCommand};
```

Comment:
I don't think splitting the subcommands facilitates review.
For this type of change, review works best by comparing corresponding lines. Unlike review time for actual logic changes, this scales linearly.

I'd keep everything in one file but if we decide otherwise we should squash this commit to minimize code movement.

Reply:
Yeah I find this splitting excessive.

Having files with 50 lines just means you need to open another file just to see the 50 lines.

I don't believe the ~2000 lines of string.cpp have been an issue for me in navigation, and these commands are more similar than they are different.

Reply:
While I agree it can see excessive, my point here was originally to have one-option-struct-per subcommand, and split up the intertwined argument-parsing. That I believe we want regardless.

Each subcommand is independent and does its own thing, so they don't need to be in the same file, as there is little logic being shared between them. While I agree some of these end up being very short, others, like `replace` and `match` are not, and I err on the side of consistency than only splitting those or splitting e.g. `escape` and `unescape` into the same module. Unlike most other builtins you never use the bare `string` builtin, so this setup makes more sense.

Also: I don't think this is that important. If we are keeping the splitting of options to one-per-subcmd this code is not directly 1-1 with the C++ version anyways, so I don't see the point spending time merging it back together and rebasing commits around it.

My thinking at the moment of the split was that it would be easier to mark one subcommand as "reviewed" — but as you point out, depending on how you perform the review, that does not necessarily actually help. I still think we want to keep this layout as it makes it clearer how independent each sub-command is, and review-ability was never my only argument for splitting it up, though the commit-message did not reflect that.

Reply:
> While I agree it can see excessive, my point here was originally to have one-option-struct-per subcommand, and split up the intertwined argument-parsing. That I believe we want regardless.

definitely

> Each subcommand is independent and does its own thing, so they don't need to be in the same file, as there is little logic being shared between them. While I agree some of these end up being very short, others, like `replace` and `match` are not, and I err on the side of consistency than only splitting those or splitting e.g. `escape` and `unescape` into the same module. Unlike most other builtins you never use the bare `string` builtin, so this setup makes more sense.

That's a good point, the implementation details of each subcommand need not be visible elsewhere.

> Also: I don't think this is that important. If we are keeping the splitting of options to one-per-subcmd this code is not directly 1-1 with the C++ version anyways, so I don't see the point spending time merging it back together and rebasing commits around it.

I agree this one is maybe not terribly important but we should generally prioritize readability over other, shorter-termed considerations.
Since there are only two logical commits touching the string builtin, it shouldn't take more than 20 minutes to merge the subcmds back together.

> I still think we want to keep this layout as it makes it clearer how independent each sub-command is, and review-ability was never my only argument for splitting it up, though the commit-message did not reflect that.

The independence is a fair point. OTOH single-file also has advantages; it's more consistent with other builtins, avoids noise in the file tree and makes it a bit easier to spot common patterns.

I don't have a super strong opinion on whether to split or not, but how about a compromise:
since we're not confined to Java visibility conventions, we can put them all in the same file while still making implementations private:

```rust
mod collect {
	pub struct Collect {
	}
	fn collect_impl() { ... }
}
mod escape {
	pub struct Escape {
	}
	...
}
```

(The extra indentation is not great..)

Reply:
> OTOH single-file also has advantages; it's more consistent with other builtins, avoids noise in the file tree and makes it a bit easier to spot common patterns.

While it one the surface might be more consistent with other builtins, `string` is ~2x larger than any other builtin, and I therefore think it is more than fine for it to be an exception. Just look at https://fishshell.com/docs/current/commands.html#the-full-list, each subcommand has its own documentation link, unlike all other builtins. And unlike C++ we don't have to manually add each file to `CMakeLists.txt`.

I just think the general setup of

```
cmd.rs
cmd/
├── subcmd1.rs
├── subcmd2.rs
├── subcmd3.rs
├── ...
```

Just makes _so much sense_

Aside: I think we should aim for `StringSubCommand` and `Arguments` to be re-usable from other builtins, which for `Arguments` should just be a couple (const) generic params away and a `pub(self) type Arguments<'args, 'iter> = super::shared::Arguments<'args, 'iter, X, Y>;` away, but that's for when `path` gets ported.

> (The extra indentation is not great..)

Rustfmt in general introduces a lot of indentation, especially with `match`-statements. The code is already at least one level more indented than the C++ code due to it being in `impl`-blocks. For consistency I don't want to have just the commands that declare a helper-struct in their own modules, and I think we both agree putting the full helper-struct decl and impl to _within_ the `handle` is an even worse solution. The act of encapsulation within the string module is not that important, so I'd rather have it all in the same module if it's in a single file, as additional indentation is worse.

Reply:
I don't see a strong argument against separate files so it's fine